In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

Penjelasan:
Pada tahap pertama, beberapa library yang dibutuhkan diimpor terlebih dahulu. Library NumPy digunakan untuk mengolah data dalam bentuk array, sedangkan Pandas digunakan untuk membuat dan mengolah data dalam bentuk DataFrame. Selain itu, `load_iris` dari Scikit-learn digunakan untuk mengambil dataset Iris yang sudah tersedia di dalam library Scikit-learn.
Dataset Iris nantinya digunakan sebagai data untuk melakukan klasifikasi menggunakan algoritma Naïve Bayes.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Penjelasan: Mengubah Dataset Menjadi DataFrame dan Menentukan Fitur serta Target
Pada tahap ini data dari dataset Iris diubah menjadi DataFrame menggunakan Pandas agar lebih mudah untuk dilihat dan diolah.
`iris.data` berisi nilai dari empat fitur pada bunga Iris, sedangkan `iris.feature_names` digunakan sebagai nama kolomnya. Setelah itu dibuat kolom baru bernama `Species` yang berisi nama species dari setiap data bunga Iris.

Selanjutnya data dipisahkan menjadi dua bagian. Variabel `X` digunakan sebagai fitur atau data masukan yang terdiri dari empat karakteristik bunga, yaitu sepal length, sepal width, petal length, dan petal width. Sedangkan variabel `y` digunakan sebagai target atau label yang berisi species bunga.
Kolom `Species` tidak dimasukkan ke dalam `X` karena kolom tersebut merupakan hasil atau kelas yang nantinya akan diprediksi oleh model.


In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Penjelasan:

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Penjelasan:Mengubah Label Species Menjadi Angka

Pada tahap ini label species yang sebelumnya masih berupa teks diubah menjadi bentuk angka menggunakan `LabelEncoder`.
Sebelumnya, nilai pada `y` masih berupa nama species, yaitu `setosa`, `versicolor`, dan `virginica`. Agar dapat digunakan dalam proses klasifikasi, label tersebut diubah menjadi angka.
`LabelEncoder()` akan memberikan kode angka untuk setiap kelas. Setelah proses `fit_transform()` dilakukan, label species menjadi nilai numerik, yaitu 0, 1, dan 2. Nilai tersebut tetap mewakili species yang sama, hanya bentuk datanya saja yang diubah menjadi angka.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Penjelasan: Membagi Data Training dan Data Testing

Pada tahap ini dataset dibagi menjadi data training dan data testing menggunakan `train_test_split()`.
Data training digunakan untuk melatih model Naïve Bayes agar model dapat mempelajari pola dari data yang tersedia. Sedangkan data testing digunakan untuk menguji hasil prediksi model menggunakan data yang tidak digunakan saat proses pelatihan.
Parameter `test_size=0.2` berarti 20% data digunakan sebagai data testing dan 80% sisanya digunakan sebagai data training. `random_state=0` digunakan agar pembagian data tetap sama ketika kode dijalankan kembali.
Pembagian dilakukan untuk `X` dan `y` sehingga data fitur dan labelnya tetap digunakan sesuai dengan kebutuhan model.

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

Penjelasan: Membuat dan Melatih Model Gaussian Naïve Bayes

Pada tahap ini digunakan algoritma Gaussian Naïve Bayes untuk melakukan klasifikasi dataset Iris.

Gaussian Naïve Bayes digunakan karena fitur pada dataset Iris berupa data numerik kontinu. Model dibuat menggunakan `GaussianNB()` kemudian dilatih menggunakan data training yang sudah dibuat sebelumnya.

`model.fit(X_train, y_train)` digunakan untuk melatih model berdasarkan data fitur dan label pada data training. Setelah model selesai dilatih, `predict(X_test)` digunakan untuk melakukan prediksi terhadap data testing.


In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Penjelasan:


In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Penjelasan: Menghitung Accuracy Model

Setelah model melakukan prediksi terhadap data testing, langkah berikutnya adalah menghitung tingkat akurasi dari hasil prediksi tersebut.
Pada bagian ini digunakan `accuracy_score()` untuk membandingkan label sebenarnya pada `y_test` dengan hasil prediksi yang terdapat pada `result`.

Hasil akurasi kemudian dikalikan dengan 100 agar ditampilkan dalam bentuk persentase. Dari hasil pengujian pada modul, diperoleh accuracy sebesar 96.66666666666667%, sehingga sebagian besar data testing berhasil diklasifikasikan dengan benar oleh model.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


C:\Users\vivob\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Penjelasan:  Melakukan Prediksi Data Baru

Setelah model selesai dilatih dan diuji, model kemudian digunakan untuk melakukan prediksi terhadap satu data baru.
Pada contoh di modul, data baru yang digunakan adalah `[5.1, 3.5, 1.4, 0.2]`. Data tersebut berisi empat nilai yang sesuai dengan empat fitur pada dataset Iris, yaitu sepal length, sepal width, petal length, dan petal width.
Data baru kemudian dimasukkan ke dalam `NB_model.predict()` untuk mengetahui kelas atau species yang diprediksi oleh model.

Karena hasil prediksi dari model masih berupa angka, digunakan `lab.inverse_transform()` untuk mengubah kembali angka tersebut menjadi nama species agar hasilnya lebih mudah dipahami.

In [ ]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Penjelasan:  Prediksi Data Baru Menggunakan DataFrame

Pada tahap ini data input yang sebelumnya berbentuk array diubah menjadi DataFrame. Empat nilai input diberikan nama kolom yang sama dengan fitur pada dataset Iris.
Setelah data dibuat dalam bentuk DataFrame, data tersebut dimasukkan ke dalam model `NB_model` untuk mendapatkan hasil prediksi.

Hasil prediksi yang masih berupa angka kemudian dikembalikan ke nama species menggunakan `inverse_transform()`. Dengan cara ini, hasil prediksi dapat langsung ditampilkan dalam bentuk nama species.

Tugas

1. Ikuti semua tahapan dalam modul ini.
2. Prediksi data baru masuk ke species apa: [7.3, 3.8, 4.6, 2.3]

In [ ]:
input_data = pd.DataFrame(
    np.array([[7.3, 3.8, 4.6, 2.3]]),
    columns=iris.feature_names)
    
input_data

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,7.3,3.8,4.6,2.3


Penjelasan:  Hasil Input Data

Data baru berhasil dibuat dalam bentuk DataFrame dengan empat fitur yang sama seperti dataset Iris. Nilai yang digunakan adalah 7.3 untuk sepal length, 3.8 untuk sepal width, 4.6 untuk petal length, dan 2.3 untuk petal width.

Data tersebut kemudian akan digunakan sebagai input untuk model Gaussian Naïve Bayes yang sudah dilatih sebelumnya.

In [12]:
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)

print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['virginica']


Penjelasan: Hasil Prediksi

Berdasarkan hasil prediksi menggunakan model Gaussian Naïve Bayes, data baru dengan nilai `[7.3, 3.8, 4.6, 2.3]` diprediksi masuk ke dalam species **virginica**.
Hasil tersebut diperoleh dari model yang sebelumnya sudah dilatih menggunakan dataset Iris. Nilai hasil prediksi dari model kemudian diubah kembali menjadi nama species menggunakan `inverse_transform()` sehingga hasil akhirnya dapat ditampilkan sebagai nama species.


*Kesimpulan:*
Berdasarkan proses klasifikasi yang telah dilakukan menggunakan algoritma Gaussian Naïve Bayes pada dataset Iris, model berhasil digunakan untuk melakukan prediksi species bunga. Pada data baru `[7.3, 3.8, 4.6, 2.3]`, hasil prediksi menunjukkan bahwa data tersebut termasuk ke dalam species **virginica**.